### PR TITLE
JAMES-1736 Introduce MessageParser which retrieve the list of attachm…

### PR DESCRIPTION
**Pull request by aduprat**
…ents of a message

### CODE REVIEW

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        List<Attachment> attachments = testee.retrieveAttachments(ClassLoader.getSystemResourceAsStream("eml/embeddedAttachmentWithInline.eml"));
+
+        assertThat(attachments).hasSize(1);
+    }
+
+    @Test
+    public void getAttachmentsShouldNotRetrieveInlineAttachmentsWhenSome() throws Exception {
+        List<Attachment> attachments = testee.retrieveAttachments(ClassLoader.getSystemResourceAsStream("eml/embeddedAttachmentWithAttachment.eml"));
+
+        assertThat(attachments).hasSize(1);
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java]: Can't you make any asserts on the attachement content ?

**Comment by rouazana** [comment]
:+1:

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java (hunk 1 above)
Comment (aduprat) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java]: That will be difficult:
- mock
- having the attachment in another file

And it looks like testing mime4j-dom

**Comment by chibenwa** [comment]
:+1:

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java (hunk 1 above)
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java]: ok

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+public class Attachment {
+
+    private final InputStream stream;
+
+    public Attachment(InputStream stream) {
+        this.stream = stream;
+    }
+
+    public InputStream getStream() {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java]: I find this API confusing : you won't be able to consume this stream twice, this objects change its state when something else read its stream.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+import com.google.common.collect.ImmutableList;
+
+public class MessageParser {
+
+    private final MessageBuilder messageBuilder;
+    private final MessageWriter messageWriter;
+
+    public MessageParser() {
+        MessageServiceFactoryImpl messageServiceFactoryImpl = new MessageServiceFactoryImpl();
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: what makes you think you need to create these three objects ahead of using them ?

Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: In order to simplify this task, as we don't know yet its future usage; i doesn't want to introduce a service but rather an instanciable class.
Then its responsibility is to parse a message, which will be done thru a MessageBuilder and a MessageWriter; so it's seems logical to  bootstrap them.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java (hunk 2 above)
Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java]: Right

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java [hunk 4]
```diff
@@ -29,7 +33,7 @@ public Attachment(InputStream stream) {
         this.stream = stream;
     }
 
-    public InputStream getStream() {
-        return stream;
+    public InputStream getStream() throws IOException {
+        return new ByteArrayInputStream(IOUtils.toByteArray(stream));
```
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java]: I think you still have an issue, as the stream field can be consumed only once.
Can you add a test about this?

Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java]: Ok, but then we may have to consume the inputStream, and then what is the added value of keeping an InputStream in the API?
We may use a byte array, WDYT?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        Body body = messageBuilder.parseMessage(fullContent)
+                .getBody();
+        if (body instanceof Multipart) {
+            Multipart multipart = (Multipart) body;
+            for (Entity entity : multipart.getBodyParts()) {
+                if (isAttachmentEntity(entity)) {
+                    attachments.add(createAttachment(entity.getBody()));
+                }
+            }
+            body.dispose();
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: you don't dispose if it's not multipart ?

**Comment by aduprat** [comment]
test this please

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        messageWriter = messageServiceFactoryImpl.newMessageWriter();
+    }
+
+    public List<Attachment> retrieveAttachments(InputStream fullContent) throws MimeException, IOException {
+        ImmutableList.Builder<Attachment> attachments = ImmutableList.builder();
+        Body body = messageBuilder.parseMessage(fullContent)
+                .getBody();
+        if (body instanceof Multipart) {
+            Multipart multipart = (Multipart) body;
+            for (Entity entity : multipart.getBodyParts()) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: too much code in this method, you should extract some things and avoid nested calls

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+            }
+            body.dispose();
+        }
+        return attachments.build();
+    }
+    
+    private Attachment createAttachment(Body body) throws IOException {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        messageWriter.writeBody(body, out);
+        body.dispose();
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: you should dispose at the same place you create the entity

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        return new Attachment(new ByteArrayInputStream(out.toByteArray()));
+    }
+
+    private boolean isAttachmentEntity(Entity part) {
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: isAttachment seems enough

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+
+import org.apache.james.mailbox.store.mail.model.Attachment;
+import org.junit.Before;
+import org.junit.Test;
+
+public class MessageParserTest {
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java]: we only have that many tests in opush ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 3 above)
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: I understand you want to instantiate it vs defining a singleton but I don't see why we don't create a builder when we need one. What do we gain with this premature instantiation ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java (hunk 4 above)
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java]: it's only a matter of naming IMO.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 3 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: ok, reading some code examples we don't need MessageServiceFactoryImpl at all and we can instanciate default where they are used.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java (hunk 4 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java]: what do you suggest ? getOneShotInputStream?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 5 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: fixed

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 6 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: fixed

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 7 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: ok

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java (hunk 8 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java]: ok

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java (hunk 9 above)
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/impl/MessageParserTest.java]: nope, but only these one are interesting for what we want to do. We have also added one test (true embedded email with embedded attachment).
Other tests are only testing mime4j...

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java (hunk 4 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/Attachment.java]: In fact already fixed.

**Comment by rouazana** [comment]
Comments fixed in #235

**Pull request closed without merging**